python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+    # As if calling register_activity on each
+    worker.register_activities([another_activity, MyCallableClass()])
+    # As if calling register_activity with a custom name on each
+    worker.register_activities({'activity_name': simple_activity})
+
+    # Fails, has no @activity.run method
+    worker.register_activities(MyCallableClass())
+```
+
+Notes:
+
+* We use `@activity.run` from `temporalio.activity` instead of just `@activity` from because:
+  * We want `activity` to be a module because we want it usually imported as `from temporalio import activity` so
+    top-level calls like `activity.info()` read well
+  * When we get to workflows, we will do similar with `@workflow.run` vs `@workflow.signal` vs `@workflow.query`, etc
+    and `workflow.info()`
+* Cancellation is task cancellation
+  * Do we need to differentiate worker shutdown from server-side cancellation?
+  * So we need an explicit future they can wait on for activity cancellation?
+
+### Data Converters
+
+Exported from `temporalio` (but probably in `temporalio/converter/converter.py`, exported from `temporalio.converter`
+then exported from `temporalio`):
+
+```python
+class DataConverter(ABC):
+    # This can be sync or async. Always takes a list. Can return a single
+    # payload or a list of payloads with the same size as the input list or None
+    # if the value cannot be converted with this converter. Any other result
+    # causes an error. Any error raised is bubbled.
+    @abstractmethod
+    async def encode(self, *values):
+        pass
+
+    # This can be sync or async. Always takes a list. Can return a single value
+    # or a list of values with the same size as the input list or None if the
+    # value cannot be converted with this converter. Any other result causes an
+    # error. Any error raised is bubbled.
+    @abstractmethod
+    async def decode(self, *payloads):
+        pass
+
+
+class CompositeDataConverter(DataConverter):
+    def __init__(self, converters) -> None:
+        pass
+
+    async def encode(self, *values):
+        pass
+
+    async def decode(self, *payloads):
+        pass
+
+
+DataConverter.default = CompositeDataConverter(
+    [
+        BinaryNullDataConverter(),
+        BinaryPlainDataConverter(),
+        JSONProtoDataConverter(),
+        BinaryProtoDataConverter(),
+        JSONPlainDataConverter(),
+    ]
+)
+```
+
+Notes:
+
+* May be a bit difficult to allow async and non-async converter w/ same base but we can
+  * Ref https://stackoverflow.com/questions/47555934/how-require-that-an-abstract-method-is-a-coroutine
+* This is a combination of the otherwise separate concepts of "DataConverter" and "PayloadConverter" from other SDKs
+  since the extra abstraction seems unnecessary, is that ok?
+
+### Other API Items
+
+The following API items aren't in the proposal because their use should be straightforward:
+
+* We'll extend exception and properly convert to/from for most errors at
```

Comment:
> convert to/from

I should clarify that we will keep raw protos available which is important for failure details

Reply:
Good use for `raise ... from` https://docs.python.org/3/tutorial/errors.html#exception-chaining

Reply:
We'd also need all of the `TemporalFailure` subclasses (https://github.com/temporalio/sdk-typescript/blob/690ea3c1ca0887b957b5b672689489daa6b77f55/packages/common/src/failure.ts#L48)

Reply:
> Good use for raise ... from

Yes, the example inside the activity uses `raise` `from`. This is a different deal, this is saying I need to keep the raw proto available inside the exception class

> We'd also need all of the TemporalFailure subclasses

That's what I mean with this line you are commenting on when saying:

> We'll extend exception and properly convert to/from for most errors at https://pkg.go.dev/go.temporal.io/api/serviceerror